editor/src/components/canvas/controls/observer-hooks.ts

```diff
@@ -0,0 +1,82 @@
+import React from 'react'
+import { ElementPath } from '../../../core/shared/project-file-types'
+import { fastForEach } from '../../../core/shared/utils'
+import * as EP from '../../../core/shared/element-path'
+import { CanvasRectangle, windowPoint } from '../../../core/shared/math-utils'
+import { windowToCanvasCoordinatesGlobal } from '../dom-lookup'
+
+function attachObserverToScrollingElement(
+  observerRef: React.MutableRefObject<MutationObserver | null | undefined>,
+) {
+  // TODO FIND SOMETHING NICER
+  // TODO use HTMLElement that is changing attributes on scroll (instead of cssvar) specifically for these controls
+  const thisElementRerendersOnScroll = document.getElementById('node-connectors')
+  if (thisElementRerendersOnScroll != null && observerRef.current != null) {
+    observerRef.current.observe(thisElementRerendersOnScroll, {
+      attributes: true,
+    })
+  }
+}
+
+function findTargetHtmlElement(path: ElementPath): HTMLElement | null {
+  return document.querySelector(`*[data-paths~="${EP.toString(path)}"]`)
+}
+
+export function findFramesFromDOM(targets: Array<ElementPath>): Array<CanvasRectangle> {
+  const frames: Array<CanvasRectangle> = []
```

Comment:
```suggestion
  let frames: Array<CanvasRectangle> = []
```
To highlight this variable is being mutated